### PR TITLE
Support extended mode on projective transform

### LINKED ISSUE #1318: tfa.image.rotate Different Fill Modes
**Describe the feature and the current behavior/state.**
Currently, when an image is rotated using the `tfa.image.rotate()` function, the outer regions of the image are filled with 0's, resulting in an image with black corners usually. This is usually not preferred, especially when the user is trying to create new images via data augmentation. Images with black corners are usually not a good representation of the intended creation.

`tf.keras.preprocessing.image.random_rotation` has a really good option where you can control how exactly the blank space is filled. The following pictures shows some of the common methods and how the resulting image looks closer to the original.

It would be nice to see the same option added to tfa.image.rotate.

**Relevant information**
- Are you willing to contribute it (yes/no): Yes!
- Are you willing to maintain it going forward? (yes/no): Yes!
- Is there a relevant academic paper? (if so, where): No
- Is there already an implementation in another framework? (if so, where): Keras (obviously)
- Was it part of tf.contrib? (if so, where): Not that I am aware of

**Which API type would this fall under (layer, metric, optimizer, etc.)**
`tfa.image`

**Who will benefit with this feature?**
Anyone who is currently using `tfa.image.rotate` for data / image augmentation purposes and want different methods to fill the image boundaries to create better augmented images for training.

**Any other info.**
While `tf.keras.preprocessing.image.random_rotation` (assuming you are performing rotations randomly) is capable of doing the same thing, it is generally not convenient to use it in a data pipeline. For one reason, it's not really designed for tensors in the form of `(num_pictures, num_rows, num_cols, num_channels)`, the channels last format which is whats normally expected. Secondly, it expects inputs as Numpy arrays. Finally it is hard to use directly with @tf.function. Thus, I think a native implementation would be preferred.

### PR DESCRIPTION
**Pull request by WindQAQ**
Closed #1318. Support extend mode on projective transform. Mainly follow the [scipy's implementation](https://github.com/scipy/scipy/blob/master/scipy/ndimage/src/ni_interpolation.c).

From top to bottom: original, reflect, constant, mirror, nearest, wrap. Addons is on the left, scipy is on the right.

test

Testing script:
```python3
import numpy as np

import tensorflow as tf
import tensorflow_addons as tfa

from scipy import ndimage, misc

image = misc.face()

tf.io.write_file("test.jpeg", tf.io.encode_jpeg(image))

for mode in ["CONSTANT", "REFLECT", "WRAP", "MIRROR", "NEAREST"]:
    t = tfa.image.rotate(image, np.pi / 4, extend=mode)
    tf.io.write_file(mode.lower() + ".jpeg", tf.io.encode_jpeg(t))
    st = ndimage.rotate(image, 45, mode=mode.lower(), reshape=False)
    tf.io.write_file("scipy-" + mode.lower() + ".jpeg", tf.io.encode_jpeg(st))
```

### CODE REVIEW

**Comment by AakashKumarNain** [comment]
@WindQAQ overall looking good but I have one query. Please check the comments

File: tensorflow_addons/custom_ops/image/cc/kernels/image_projective_transform_op.cc [hunk 1]
```diff
@@ -43,32 +43,70 @@ template struct FillProjectiveTransform<CPUDevice, double>;
 
 }  // end namespace functor
 
-typedef Eigen::ThreadPoolDevice CPUDevice;
+using CPUDevice = Eigen::ThreadPoolDevice;
```
Comment (AakashKumarNain) [inline tensorflow_addons/custom_ops/image/cc/kernels/image_projective_transform_op.cc:46]: Are we forcing to run all these ops on CPU only?

Comment (WindQAQ) [reply tensorflow_addons/custom_ops/image/cc/kernels/image_projective_transform_op.cc:46]: Nope, GPUDevice is also registered [here](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/custom_ops/image/cc/kernels/image_projective_transform_op.cc#L148-L160).

**Comment by WindQAQ** [comment]
Just figure out that there is another PR #1719. @srilman what do you think about this :-) Thank you.

File: tensorflow_addons/custom_ops/image/cc/kernels/image_projective_transform_op.cc (hunk 1 above)
Comment (AakashKumarNain) [reply tensorflow_addons/custom_ops/image/cc/kernels/image_projective_transform_op.cc:46]: Cool. Thanks for the clarification

**Comment by srilman** [comment]
@WindQAQ Looks great to me! Definitely preferable to #1719. Only thing is that it would be nice if the different extended modes could be used separately (outside of projective transforms) for more flexibility.

For example, I mentioned that I liked to extend the image before performing some image ops like `tf.image.random_crop`. In general, extend can be used in lieu of `tf.image.resize_with_pad` in nearly every situation.

**Comment by WindQAQ** [comment]
> @WindQAQ Looks great to me! Definitely preferable to #1719. Only thing is that it would be nice if the different extended modes could be used separately (outside of projective transforms) for more flexibility.
> 
> For example, I mentioned that I liked to extend the image before performing some image ops like `tf.image.random_crop`. In general, extend can be used in lieu of `tf.image.resize_with_pad` in nearly every situation.

We can have both of them actually if they are both useful :-)

**Comment by bhack** [comment]
I have a question but I don't know if it is strictly related. Do you know why we have our own  custom `ImageProjectiveTransformV2`? What is https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/image_ops.cc#L52?

**Comment by WindQAQ** [comment]
@bhack It seems that it is migrated to core TF in Jan. I have no idea about this. Thanks for pointing out!

cc @tensorflow/sig-addons-maintainers 

https://github.com/tensorflow/tensorflow/commit/26992815bf7ee136c49f658596a9fbde7881d425

**Comment by bhack** [comment]
@WindQAQ I know that I am blamed as the cross-linking guy :stuck_out_tongue_winking_eye:  but I suppose it was part of a more general effort we already commented at https://github.com/tensorflow/addons/issues/1226#issuecomment-608594527 and some follow-up messages in that thread mentioning Keras.

**Comment by WindQAQ** [comment]
> @WindQAQ I know that I am blamed as the cross-linking guy 

Never mind :smile: I don't even care about it LOL. Your links are super helpful for me.

> but I suppose it was part of a more general effort that we already commented at [#1226 (comment)](https://github.com/tensorflow/addons/issues/1226#issuecomment-608594527) and in some follow-up messages in that thread mentioning Keras.

Umm, it seems to be the miscommunication between us and many places. A related issue has been posted https://github.com/tensorflow/addons/issues/1779. Thanks again for pointing out it :-)

**Comment by bhack** [comment]
@WindQAQ Users are looking for some of these filling parameters at https://github.com/tensorflow/tensorflow/issues/39177
